File: packages/ruby/lib/uncaught/rate_limiter.rb
# frozen_string_literal: true

module Uncaught
  # Sliding-window rate limiter.
  #
  # Enforces:
  #   - A global maximum of events per 60-second window.
  #   - A per-fingerprint maximum of events per 60-second window.
  #
  # Thread-safe via Mutex.
  class RateLimiter
    # Window duration in seconds.
    WINDOW_SECONDS = 60

    # @param global_max          [Integer] Max events across all fingerprints. Defaults to 30.
    # @param per_fingerprint_max [Integer] Max events for a single fingerprint. Defaults to 5.
    def initialize(global_max: 30, per_fingerprint_max: 5)
      @global_max = global_max
      @per_fingerprint_max = per_fingerprint_max
      @global_timestamps = []
      @buckets = {}
      @mutex = Mutex.new
    end

    # Returns true if the event identified by fingerprint is allowed through,
    # false if it should be dropped.
    #
    # @param fingerprint [String]
    # @return [Boolean]
    def should_allow?(fingerprint)
      @mutex.synchronize do
        now = Time.now.to_f

        # --- Global limit ---
        prune!(@global_timestamps, now)
        return false if @global_timestamps.size >= @global_max

        # --- Per-fingerprint limit ---
        @buckets[fingerprint] ||= []
        fp_timestamps = @buckets[fingerprint]
        prune!(fp_timestamps, now)
        return false if fp_timestamps.size >= @per_fingerprint_max

        # Record this event
        @global_timestamps << now
        fp_timestamps << now

        # Periodic cleanup: remove empty buckets
        if @buckets.size > 1000
          @buckets.delete_if { |_k, v| v.empty? }
        end

        true
      end
    end

    private

    # Remove timestamps older than the sliding window from an array (in-place).
    def prune!(timestamps, now)
      cutoff = now - WINDOW_SECONDS
      timestamps.shift while timestamps.any? && timestamps.first <= cutoff
    end
  end
end

File: packages/dotnet/src/Uncaught/RateLimiter.cs
// ---------------------------------------------------------------------------
// Uncaught — sliding-window rate limiter (thread-safe)
// ---------------------------------------------------------------------------

namespace Uncaught;

/// <summary>
/// Sliding-window rate limiter that enforces:
///  - A global maximum of events per 60-second window.
///  - A per-fingerprint maximum of events per 60-second window.
/// </summary>
public sealed class RateLimiter
{
    private const long WindowMs = 60_000;

    private readonly int _globalMax;
    private readonly int _perFingerprintMax;
    private readonly object _lock = new();

    private readonly List<long> _globalTimestamps = new();
    private readonly Dictionary<string, List<long>> _buckets = new();

    public RateLimiter(int globalMax = 30, int perFingerprintMax = 5)
    {
        _globalMax = globalMax;
        _perFingerprintMax = perFingerprintMax;
    }

    /// <summary>
    /// Returns true if the event identified by fingerprint is allowed
    /// through, false if it should be dropped.
    /// </summary>
    public bool ShouldAllow(string fingerprint)
    {
        lock (_lock)
        {
            var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            var cutoff = now - WindowMs;

            // Prune and check global limit
            _globalTimestamps.RemoveAll(t => t <= cutoff);
            if (_globalTimestamps.Count >= _globalMax)
            {
                return false;
            }

            // Prune and check per-fingerprint limit
            if (!_buckets.TryGetValue(fingerprint, out var fpTimestamps))
            {
                fpTimestamps = new List<long>();
                _buckets[fingerprint] = fpTimestamps;
            }

            fpTimestamps.RemoveAll(t => t <= cutoff);
            if (fpTimestamps.Count >= _perFingerprintMax)
            {
                return false;
            }

            // Record this event
            _globalTimestamps.Add(now);
            fpTimestamps.Add(now);

            // Periodic cleanup: remove empty buckets
            if (_buckets.Count > 1000)
            {
                var emptyKeys = _buckets
                    .Where(kvp => kvp.Value.Count == 0)
                    .Select(kvp => kvp.Key)
                    .ToList();

                foreach (var key in emptyKeys)
                {
                    _buckets.Remove(key);
                }
            }

            return true;
        }
    }
}
